### DISCUSSION TITLE
Listener to react on mouse clicks on data points

### DISCUSSION THREAD

**Discussion by magicroomy** [GSI-CS-CO/chart-fx#431] [Q&A]
I am looking for an example to establish a listener to react on the user clicking on a single data point (Entry in a dataset). So I could direct the user to the "detail" of the point in another window.
If it is important: I use DoubleErrorDataSets in XY Graphs.

**Comment by wirew0rm** [comment]
Because the rendering is based on the canvas and not on the scene graph, you cannot directly get a listener to individual data points.

As the DataPointTooltip plugin solves a similar problem, you can probably use a lot of its code to create your own plugin. See [here](https://github.com/fair-acc/chart-fx/blob/5719d922926d74d0f77fe34f1cfc1311fc171118/chartfx-chart/src/main/java/io/fair_acc/chartfx/plugins/DataPointTooltip.java#L80) how the mouse listener is set up. It also has the code necessary to find the closest data point to the clicked mouse position (Not trivial, because a chart can have multiple renderers with different axes, datasets can be added to multiple renderers at the same time, datareduction).

**Reply by magicroomy** [reply]
Thank you for the hint... It worked (I almost copied the complete DataPointTooltip :-) )
And again I have to say.. great work with ChartFX !!!

**Reply by thhart** [reply]
@magicroomy Any chance to share some code here, I am looking for the same functionality?

**Reply by magicroomy** [reply]
Sorry, bad news. I had to do some software archeology. I turned out that I kicked the idea of having click events on data points, sorry.
I also tried to follow the like described by wirew0rm to maybe reconstruct it. The page is gone... 404
So sorry again.

**Reply by wirew0rm** [reply]
Replaced the link with a permalink.

**Reply by yezhengli-Mr9** [reply]
> @magicroomy Any chance to share some code here, I am looking for the same functionality?

Same question, **any code for this is welcome**~ @magicroomy 
Used to try things like following but not work so far.

- adding `Tooltip` or `DatapointTooltip` (as @wirew0rm  pointed)
https://github.com/fair-acc/chart-fx/blob/5719d922926d74d0f77fe34f1cfc1311fc171118/chartfx-chart/src/main/java/io/fair_acc/chartfx/plugins/DataPointTooltip.java#L75-L81
after 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-chart/src/main/java/io/fair_acc/chartfx/renderer/spi/financial/CandleStickRenderer.java#L215-L219
```java
                 Rectangle rect = new Rectangle(x0-barWidthHalf, yLow, x0+barWidthHalf,yHigh);
                Tooltip t = new Tooltip("Hello, World!");
                Tooltip.install(rect, t);

                rect.setOnMouseMoved(event -> {
                    if (x0-barWidthHalf< event.getX() &&
                            event.getX() <x0+barWidthHalf &&
                            yLow< event.getY() &&
                            event.getY()< yHigh
                    ) {
                        t.show(rect, event.getScreenX(), event.getScreenY());
                    } else {
                        t.hide();
                    }
                });
```

- add to   https://blog.csdn.net/qq_40946921/article/details/84858886
```java
Text text = new Text("Hover over the rectangle +("+i+","+j+")");
                cover[i][j].setOnMouseEntered(e-> {
                    cover[a][b].setFill(Color.rgb(150, 180, 200));
                    if (cover[a][b].contains(e.getX(), e.getY())) {
                        text.setVisible(true);
                    } else {
                        text.setVisible(false);
                    }
//                    getChildren().addAll(cover[a][b], text);
                });
```
- `Tooltip(...)` in https://www.geeksforgeeks.org/javafx-tooltip/

Also checking `findNearestDataPointWithinPickingDistance(...)`  in different versions (for earlier version #527 )
https://github.com/fair-acc/chart-fx/blob/cd202c7c9a0b40f1e66b6d0fe8fd93fffbec5435/chartfx-chart/src/main/java/de/gsi/chart/plugins/DataPointTooltip.java#L102-L115
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-chart/src/main/java/io/fair_acc/chartfx/plugins/DataPointTooltip.java#L103-L116
